### DISCUSSION TITLE
How can I get the request protocol/scheme?

### DISCUSSION THREAD

**Discussion by ghost** [tokio-rs/axum#858] [Q&A]
How can I get the incoming request's protocol/scheme, e.g. http, https, etc.?
Alternatively how can I get the full URI of the request starting with the protocol/scheme, e.g. http://..., https://..., erc.?
Finally, how can I do either of the above in a middleware/layer and in a request handler?

**Comment by davidpdrsn** [comment]
You can extract an `axum::http::Uri` which has all the details.

However I believe there are some limitations with [`Uri::scheme`](https://docs.rs/http/latest/http/uri/struct.Uri.html#method.scheme) and HTTPS but I'm not exactly sure. I think the easiest way is to set some configuration that tells the server if its running on HTTP or HTTPS. I suppose you know that when you start the server.

**Reply by LukeAbby** [reply]
Hey has this changed since 2022?

I cannot seem to get this to work today:
```rust
use axum::{body::Body, routing::any, Router, extract::OriginalUri};

#[tokio::main]
async fn main() {
    let app = Router::new().route("/", any(handle_root));

    axum::Server::bind(&"0.0.0.0:1234".parse().unwrap())
        .serve(app.into_make_service())
        .await
        .unwrap();
}

async fn handle_root(
    OriginalUri(original_uri): OriginalUri,
    uri: axum::http::Uri,
    request: axum::http::Request<Body>
) -> String {
    format!(
        "Hello, World! Original URI Scheme = {:?}, URI Scheme = {:?}, Request Scheme = {:?}",
        original_uri.scheme(),
        uri.scheme(),
        request.uri().scheme()
    )
}
```

Always ends up with `Hello, World! Original URI Scheme = None, URI Scheme = None, Request Scheme = None` for me. It doesn't help if you reduce the number of extractors as well.

**Reply by ttys3** [reply]
I tried the demo, got `None` for scheme too

**Reply by davidpdrsn** [reply]
Unfortunately it's not as easy as just calling `Uri::scheme`, like I first thought.

HTTP requests often don't contain the scheme, for example:

```
GET /hello.htm HTTP/1.1
User-Agent: Mozilla/4.0 (compatible; MSIE5.01; Windows NT)
Host: www.tutorialspoint.com
Accept-Language: en-us
Accept-Encoding: gzip, deflate
Connection: Keep-Alive
```

You can't tell just from the request whether it is http or https.

If you're using a proxy to handle TLS it might set an `X-*` header you can read or if you're handling TSL manually inside your app then you already know that things will be over HTTPS.

**Reply by ttys3** [reply]
all the behavior is document in https://www.rfc-editor.org/rfc/rfc7230#section-5.3

>  There are four distinct formats for the request-target,
   depending on both the method being requested and whether the request
   is to a proxy.

```
     request-target = origin-form
                    / absolute-form
                    / authority-form
                    / asterisk-form
```

```markdown
[5.3.1](https://www.rfc-editor.org/rfc/rfc7230#section-5.3.1).  origin-form

   The most common form of request-target is the origin-form.

     origin-form    = absolute-path [ "?" query ]

   When making a request directly to an origin server, other than a
   CONNECT or server-wide OPTIONS request (as detailed below), a client
   MUST send only the absolute path and query components of the target
   URI as the request-target.  If the target URI's path component is
   empty, the client MUST send "/" as the path within the origin-form of
   request-target.  A Host header field is also sent, as defined in
   [Section 5.4](https://www.rfc-editor.org/rfc/rfc7230#section-5.4).



[5.3.2](https://www.rfc-editor.org/rfc/rfc7230#section-5.3.2).  absolute-form

   When making a request to a proxy, other than a CONNECT or server-wide
   OPTIONS request (as detailed below), a client MUST send the target
   URI in absolute-form as the request-target.

     absolute-form  = absolute-URI

   The proxy is requested to either service that request from a valid
   cache, if possible, or make the same request on the client's behalf
   to either the next inbound proxy server or directly to the origin
   server indicated by the request-target.  Requirements on such
   "forwarding" of messages are defined in [Section 5.7](https://www.rfc-editor.org/rfc/rfc7230#section-5.7).

   An example absolute-form of request-line would be:

     GET http://www.example.org/pub/WWW/TheProject.html HTTP/1.1
```

you will only get `scheme` and `authority` when the client send a `absolute-form` request.

**Reply by LukeAbby** [reply]
hmmm, that seems limiting. I guess you're saying it's an intrinsic problem? I know Axum has websocket extractor but I couldn't figure out whether to proxy to `ws` or `wss` and scheme seemed like a good way. I'm sure there's other uses for like `http` versus `https`. I'm not sure how many other schemes are relevant or even possible for Axum.

**Reply by davidpdrsn** [reply]
> I guess you're saying it's an intrinsic problem?

Yes.

**Reply by Pzixel** [reply]
I wonder if it's truly an inherent problem. Actix have no problems providing this info:

```rust
use actix_web::{web, App, Responder, HttpServer, HttpRequest};

async fn handler(req: HttpRequest) -> impl Responder {
    format!("Requested path: {:?} {}", req.connection_info(), req.uri())
}

#[actix_web::main]
async fn main() -> std::io::Result<()> {
    HttpServer::new(|| App::new()
        .default_service(web::to(handler))
        )
        .bind(("127.0.0.1", 8080))?
        .run()
        .await
}
```

`connection_info` provide more detailed request info, including request schema. Response example for request with no headers:

```bash
> curl http://localhost:8080/123/456
< Requested path: ConnectionInfo { host: "localhost:8080", scheme: "http", peer_addr: Some("127.0.0.1"), realip_remote_addr: None } /123/456
```

Do you think it's not possible to do something similar in axum?

**Reply by davidpdrsn** [reply]
You're welcome to take a stab at implementing something similar. axum just forwards the data directly from the request its gets from hyper.
